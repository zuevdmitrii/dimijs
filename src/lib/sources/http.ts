import {
  CrudErrorCodes,
  CrudEvents,
  ICrudFilter,
  ICrudList,
  ICrudPagination,
  ICrudSorting,
  ICrudStatus,
  Source,
} from './source';

export class Http<ItemType, KeyFieldType> extends Source<
  ItemType,
  KeyFieldType
> {
  private customQueryParams: {[propname: string]: string} = {};
  private queryLine = '';
  // will be '${httpEndPoint}/{create|read|update|list}?'
  // create, delete and update = post methods
  // read, list = get
  public httpEndPoint: string = '/crud';
  public setCustomQueryParams = (params: {[propname: string]: string}) => {
    this.customQueryParams = params;
    this.queryLine = '';
    Object.keys(this.customQueryParams).forEach((p) => {
      if (this.queryLine) this.queryLine += '&';
      this.queryLine += `${encodeURIComponent(p)}=${encodeURIComponent(
        this.customQueryParams[p]
      )}`;
    });
    if (this.queryLine) this.queryLine = `?${this.queryLine}`;
  };
  create: (
    items: {
      [Property in keyof ItemType]+?: ItemType[Property];
    }[]
  ) => Promise<ICrudStatus> = (items) => {
    return fetch(`${this.httpEndPoint}/create${this.queryLine}`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify(items)),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((created) => {
        this.eventEmitter.fire(CrudEvents.onCreate, created);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null> = (id) => {
    return fetch(
      `${this.httpEndPoint}/read${
        this.queryLine ? `${this.queryLine}&` : '?'
      }${encodeURIComponent(String(this.keyField))}=${encodeURIComponent(
        String(id)
      )}`,
      {
        method: 'GET',
      }
    )
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property] | undefined;
  }) => Promise<ICrudStatus> = (data) => {
    return fetch(`${this.httpEndPoint}/update${this.queryLine}`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify(data)),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((updated) => {
        this.eventEmitter.fire(CrudEvents.onUpdate, [updated]);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  delete: (id: KeyFieldType) => Promise<ICrudStatus> = (id) => {
    return fetch(`${this.httpEndPoint}/delete${this.queryLine}`, {
      method: 'POST',
      body: encodeURIComponent(JSON.stringify({[this.keyField]: id})),
    })
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((deleted) => {
        if ((deleted as ICrudStatus).errorCode) return deleted;
        this.eventEmitter.fire(CrudEvents.onDelete, [deleted]);
        return {errorCode: CrudErrorCodes.OK};
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudStatus | ICrudList<ItemType>> = (
    filter,
    pagination,
    sorting
  ) => {
    return fetch(
      `${this.httpEndPoint}/list${
        this.queryLine ? `${this.queryLine}&` : '?'
      }params=${encodeURIComponent(
        JSON.stringify({filter, pagination, sorting})
      )}`,
      {
        method: 'GET',
      }
    )
      .then((response) => {
        if (response.status === 200) return response.json();
        throw new Error(response.statusText);
      })
      .then((jsonData) => {
        this.cache[JSON.stringify({filter, pagination, sorting})] = jsonData;
        return jsonData;
      })
      .catch((reason) => {
        return {errorCode: CrudErrorCodes.ERROR, errorMessage: reason.message};
      });
  };
}
