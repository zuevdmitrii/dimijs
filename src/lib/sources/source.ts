import {EventEmitter} from './eventEmitter';

export enum CrudErrorCodes {
  OK = 0,
  ERROR = 1,
}
export interface ICrudStatus {
  errorCode: CrudErrorCodes;
  errorMessage?: string;
}
export enum CrudFilterOperator {
  EQ,
  GT,
  GTE,
  LT,
  LTE,
  NOTEQ,
  LIKE,
}
export enum CrudAggregation {
  AND,
  OR,
}
export enum CrudFilterFieldTypes {
  Aggregation,
  Field,
}
export type ICrudFilterField<ItemType, T extends keyof ItemType> = {
  operator?: CrudFilterOperator;
  field: T;
  value: ItemType[T];
  type: CrudFilterFieldTypes.Field;
};

export interface ICrudAggregationFilterField<ItemType> {
  type: CrudFilterFieldTypes.Aggregation;
  aggregation: CrudAggregation;
  filter: ICrudFilter<ItemType>[];
}
export type ICrudFilter<ItemType> =
  | ICrudAggregationFilterField<ItemType>
  | ICrudFilterField<ItemType, keyof ItemType>;

export interface ICrudPagination {
  page: number;
  countOnPage: number;
}
export interface ICrudSorting<ItemType> {
  field: keyof ItemType;
  direction?: 'DESC' | 'ASC';
}
export interface ICrudList<ItemType> {
  data: ItemType[];
  meta?: {
    hasNextPage?: boolean;
  };
}
export enum CrudEvents {
  onCreate = 'onCreate',
  onDelete = 'onDelete',
  onUpdate = 'onUpdate',
}

export interface ISource<ItemType, KeyFieldType> {
  keyField: keyof ItemType;
  create: (items: ItemType[]) => Promise<ICrudStatus>;
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null>;
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property];
  }) => Promise<ICrudStatus>;
  delete: (id: KeyFieldType) => Promise<ICrudStatus>;
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudList<ItemType> | ICrudStatus>;
  // return data from local cache syncroniously for using in component render
  getSerializationData: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => ICrudList<ItemType> | ICrudStatus | null;
  on: (
    event: CrudEvents,
    callback: (event: CrudEvents, data: ItemType[]) => void
  ) => () => void;
}

export class Source<ItemType, KeyFieldType>
  implements ISource<ItemType, KeyFieldType>
{
  protected cache = {} as {
    [propname: string]: ICrudList<ItemType> | ICrudStatus;
  };
  protected eventEmitter = new EventEmitter<{
    [Property in CrudEvents]: Property;
  }>();

  public keyField: keyof ItemType;
  constructor(
    keyField: keyof ItemType,
    deserializedData?: {
      data: ICrudList<ItemType> | ICrudStatus;
      filter?: ICrudFilter<ItemType>[];
      pagination?: ICrudPagination;
      sorting?: ICrudSorting<ItemType>[];
    }
  ) {
    this.keyField = keyField;
    if (
      deserializedData &&
      deserializedData.pagination &&
      deserializedData.filter
    ) {
      const {filter, pagination, sorting} = deserializedData;
      this.cache[JSON.stringify({filter, pagination, sorting})] =
        deserializedData.data;
    }
  }
  getSerializationData: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => ICrudList<ItemType> | ICrudStatus | null = (
    filter,
    pagination,
    sorting
  ) => {
    return this.cache[JSON.stringify({filter, pagination, sorting})] || null;
  };
  on: (
    event: CrudEvents,
    callback: (event: CrudEvents, data: ItemType[]) => void
  ) => () => void = (event, callback) => {
    return this.eventEmitter.on(event, callback);
  };
  create: (items: ItemType[]) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  read: (id: KeyFieldType) => Promise<ItemType | ICrudStatus | null> = () => {
    throw new Error('Not implemeted');
  };
  update: (data: {
    [Property in keyof ItemType]+?: ItemType[Property] | undefined;
  }) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  delete: (id: KeyFieldType) => Promise<ICrudStatus> = () => {
    throw new Error('Not implemeted');
  };
  list: (
    filter: ICrudFilter<ItemType>[],
    pagination: ICrudPagination,
    sorting?: ICrudSorting<ItemType>[]
  ) => Promise<ICrudStatus | ICrudList<ItemType>> = () => {
    throw new Error('Not implemeted');
  };
}
