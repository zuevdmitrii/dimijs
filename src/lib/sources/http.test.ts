import {Http} from './http';
import {rest} from 'msw';
import {setupServer} from 'msw/node';
import {CrudErrorCodes, CrudEvents, ICrudList, ICrudStatus} from './source';

interface ITestData {
  id: number;
  title: string;
  value?: string;
}

const server = setupServer(
  rest.post('/crud/create', async (req, res, ctx) => {
    const body = JSON.parse(decodeURIComponent(await req.text()));
    if (body[0].id === 2) return res(ctx.status(500, 'duplicate id'));
    return res(ctx.json([{...body[0], value: `${body[0].title}_calculated`}]));
  }),
  rest.post('/crud/update', async (req, res, ctx) => {
    const body = JSON.parse(decodeURIComponent(await req.text()));
    if (body.id === 2) return res(ctx.status(500, 'update error'));
    return res(ctx.json({...body, value: `${body.title}_calculated`}));
  }),
  rest.post('/crud/delete', async (req, res, ctx) => {
    const body = JSON.parse(decodeURIComponent(await req.text()));
    if (body.id === 2) return res(ctx.status(500, 'row not found'));
    if (body.id === 3)
      return res(ctx.json({errorCode: 1, errorMessage: 'err msg with 200'}));
    return res(ctx.json({id: 1, title: 'test'}));
  }),
  rest.get('/crud/list', async (req, res, ctx) => {
    const params = JSON.parse(
      decodeURIComponent(String(req.url.searchParams.get('params')))
    );
    if (params.pagination.page == 1) return res(ctx.status(500, 'error list'));
    const result: ICrudList<ITestData> = {
      data: [{id: 100, title: 'test'}],
      meta: {hasNextPage: params.pagination.page === 0},
    };
    return res(ctx.json(result));
  }),
  rest.get('/crud/read', async (req, res, ctx) => {
    const id = Number(req.url.searchParams.get('id'));
    if (id === 2) return res(ctx.status(500, 'error'));
    if (id === 3) return res(ctx.json(null));
    return res(ctx.json({id: 1, title: 'test'}));
  })
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

test('create row', async () => {
  const s = new Http<ITestData, number>('id');
  let created: any;
  s.on(CrudEvents.onCreate, (event, data) => {
    created = data[0];
  });
  const res = await s.create([{id: 1, title: 'test'}]);
  expect(res.errorCode).toBe(CrudErrorCodes.OK);
  expect(created.id).toBe(1);
  expect(created.value).toBe('test_calculated');
});

test('create row error', async () => {
  const s = new Http<ITestData, number>('id');
  let created: any;
  s.on(CrudEvents.onCreate, (event, data) => {
    created = data[0];
  });
  const res = await s.create([{id: 2, title: 'test'}]);
  expect(res.errorCode).toBe(CrudErrorCodes.ERROR);
  expect(res.errorMessage).toBe('duplicate id');
  expect(created).toBe(undefined);
});

test('delete row', async () => {
  const s = new Http<ITestData, number>('id');
  let deleted: any;
  s.on(CrudEvents.onDelete, (event, data) => {
    deleted = data[0];
  });
  const res = await s.delete(1);
  expect(res.errorCode).toBe(CrudErrorCodes.OK);
  expect(deleted.id).toBe(1);
  expect(deleted.title).toBe('test');
});

test('delete row error', async () => {
  const s = new Http<ITestData, number>('id');
  let deleted: any;
  s.on(CrudEvents.onDelete, (event, data) => {
    deleted = data[0];
  });
  const res = await s.delete(2);
  expect(res.errorCode).toBe(CrudErrorCodes.ERROR);
  expect(res.errorMessage).toBe('row not found');
  expect(deleted).toBe(undefined);
  const res2 = await s.delete(3);
  expect(res2.errorCode).toBe(CrudErrorCodes.ERROR);
  expect(res2.errorMessage).toBe('err msg with 200');
  expect(deleted).toBe(undefined);
});

test('read row', async () => {
  const s = new Http<ITestData, number>('id');
  const res = await s.read(1);
  expect((res as ITestData).id).toBe(1);
  expect((res as ITestData).title).toBe('test');
});

test('read null row', async () => {
  const s = new Http<ITestData, number>('id');
  const res = await s.read(3);
  expect(res).toBeNull();
});

test('error read row', async () => {
  const s = new Http<ITestData, number>('id');
  const res = await s.read(2);
  expect((res as ICrudStatus).errorMessage).toBe('error');
});

test('update row', async () => {
  const s = new Http<ITestData, number>('id');
  let updated: any;
  s.on(CrudEvents.onUpdate, (event, data) => {
    updated = data[0];
  });
  const res = await s.update({id: 1, title: 'test'});
  expect(res.errorCode).toBe(CrudErrorCodes.OK);
  expect(updated.id).toBe(1);
  expect(updated.value).toBe('test_calculated');
});

test('update row error', async () => {
  const s = new Http<ITestData, number>('id');
  let updated: any;
  s.on(CrudEvents.onUpdate, (event, data) => {
    updated = data[0];
  });
  const res = await s.update({id: 2, title: 'test'});
  expect(res.errorCode).toBe(CrudErrorCodes.ERROR);
  expect(res.errorMessage).toBe('update error');
  expect(updated).toBe(undefined);
});

test('list', async () => {
  const s = new Http<ITestData, number>('id');
  const res = await s.list([], {page: 0, countOnPage: 10});
  expect((res as ICrudStatus).errorCode).toBe(undefined);
  expect((res as ICrudList<ITestData>).meta?.hasNextPage).toBe(true);
  expect((res as ICrudList<ITestData>).data[0].id).toBe(100);
});

test('list error', async () => {
  const s = new Http<ITestData, number>('id');
  const res = await s.list([], {page: 1, countOnPage: 10});
  expect((res as ICrudStatus).errorCode).toBe(CrudErrorCodes.ERROR);
  expect((res as ICrudStatus).errorMessage).toBe('error list');
});
