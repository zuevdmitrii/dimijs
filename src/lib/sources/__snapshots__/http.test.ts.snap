// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom params 1`] = `
Object {
  "id": 1,
  "title": "test",
}
`;
